mqt/qmap/subarchitectures.py

```diff
@@ -0,0 +1,348 @@
+        """Construct SubarchitectureOrder from qmap Architecture object."""
+        return cls.from_coupling_map(arch.coupling_map)
+
+    @classmethod
+    def from_library(cls, path: pathlib.Path) -> SubarchitectureOrder:
+        """Construct SubarchitectureOrder from stored library."""
+        temp = None
+        with pathlib.Path(path).open("rb") as f:
+            temp = pickle.load(f)
+
+        so = SubarchitectureOrder()
+        so.__dict__.update(temp.__dict__)
+
+        return so
+
+    @classmethod
+    def from_string(cls, path: str) -> SubarchitectureOrder:
+        """Construct SubarchitectureOrder from library name."""
+        if path in precomputed_backends:
+            ref = resources.files("mqt.qmap") / "libs" / (path + ".pickle")
+            with resources.as_file(ref) as lib_path:
+                return cls.from_library(lib_path)
+        return SubarchitectureOrder()
+
+    def optimal_candidates(self, nqubits: int) -> list[rx.PyGraph]:
+        """Return optimal subarchitecture candidate.
+
+        nqubits : int
+            size of circuit for which the optimal candidate should be given.
+        """
+        if nqubits <= 0 or nqubits > self.arch.num_nodes():
+            raise ValueError(
+                "Number of qubits must not be smaller or equal 0 or larger then number of physical qubits of architecture."
+            )
+
+        if nqubits == self.arch.num_nodes():
+            return [self.arch]
+
+        cands = self.__cand(nqubits)
+        trans_ord = self.__transitive_closure(self.subarch_order)
+        ref_ord = self.__reflexive_closure(trans_ord)
+
+        opt_cands = set(ref_ord[next(iter(cands))])
+        for cand in cands:
+            opt_cands = opt_cands.intersection(set(ref_ord[cand]))
+
+        ordered_cands = list(opt_cands)
+        ordered_cands.sort()
+        for cand in ordered_cands:
+            opt_cands = opt_cands.difference(trans_ord[cand])
+
+        return [self.sgs[n][i] for (n, i) in opt_cands]
+
+    def covering(self, nqubits: int, size: int) -> list[rx.PyGraph]:
+        """
+        Return covering for nqubit circuits.
+
+        The size of the covering is limited by size.
+        Note that a smaller covering might be found.
+        """
+        cov = self.__cand(nqubits)
+        po_trans = self.__transitive_closure(self.subarch_order)
+        ref_trans_po = self.__reflexive_closure(po_trans)
+        queue = list({el for cand in cov for el in ref_trans_po[cand]})
+        queue.sort(reverse=True)
+
+        po_inv = self.__inverse_relation(po_trans)
+
+        while len(cov) > size:
+            d = queue.pop()
+            cov_d = cov.intersection(po_inv[d])
+            if len(cov_d) > 1:
+                cov = cov.difference(cov_d)
+                cov.add(d)
+
+        return [self.sgs[n][i] for n, i in cov]
+
+    def store_library(self, lib_name: str | pathlib.Path) -> None:
+        """Store ordering."""
+        if type(lib_name) is str:
```

Comment:
Also not 100% sure here, but isn't the canonical way to check types `if isinstance(lib_name, str)`?

Reply:
You are right of course. Gonna fix that.